### PR TITLE
56 render react, react testing library and vue elements

### PR DESCRIPTION
**Pull request by dakebl**
This PR attempts to address #56.
It adds the ability for `axe()` to convert react elements to strings internally and render react-testing-library elements directly.

### CODE REVIEW

**Comment by citypaul** [comment]
Hi guys,

Is there any update on this one? I'm having the issue mentioned in your original post here: https://github.com/nickcolley/jest-axe/issues/56

@dakebl it looks like there's a failure on CI here. Do you want some help on this, as I'm sure it's an issue impacting many people at the moment.

**Comment by dakebl** [comment]
Thanks for the comment @citypaul, I'm not sure why it's failing in a Boron build and not Carbon.
Looking into it now but happy for the assist

File: package.json [hunk 1]
```diff
@@ -26,17 +26,21 @@
...
+    "react-dom": "16.7.0",
+    "@testing-library/react": "^8.0.5"
   },
   "jest": {
     "transform": {
       "^.+\\.js$": "async-to-gen"
-    }
+    },
+    "transformIgnorePatterns": [
+      "node_modules/(?!(@testing-library/react)/)"
```
Comment (dakebl) [inline package.json]: This was the offending line @citypaul. I forgot to update the ignore pattern to coincide with the new naming convention for react testing library:
`react-testing-library` became `@testing-library/react`

**Comment by citypaul** [comment]
Hey, you fixed it! Awesome!! 🎉

**Comment by dakebl** [comment]
> Hey, you fixed it! Awesome!! 🎉

Yeah 😄 Thanks for the spot!

File: package.json (hunk 1 above)
Comment (citypaul) [reply package.json]: Ah, this is fantastic! Thank you for this. There must be many others experiencing the same issue, so this is great!

**Comment by citypaul** [comment]
@nickcolley Any chance this could get merged? It'd make a big difference to people using react testing library 😄

**Comment by citypaul** [comment]
@dakebl one other little comment - you might wanna update the main README to include an example using RTL with the DOM node instead of the string 👍

**Comment by dakebl** [comment]
> @dakebl one other little comment - you might wanna update the main README to include an example using RTL with the DOM node instead of the string 👍

Good call 💯

**Comment by nickcolley** [comment]
Hey sorry friends, I've been burnt out so turned off all notifications.

I'll try to review this soon now!

**Comment by nickcolley** [comment]
Had a quick look, I'm not a huge fan of making this library dependant on React but if a lot of people are using React with `jest-axe` and you and @citypaul think this'll be a big improvement I think we should get it merged.

This would be a feature release not a breaking release, correct?

File: package.json [hunk 2]
```diff
@@ -26,17 +26,21 @@
...
     "jest-matcher-utils": "24.0.0",
-    "lodash.merge": "4.6.1"
+    "lodash.merge": "^4.6.2"
   },
   "devDependencies": {
     "async-to-gen": "1.4.0",
     "jest": "24.1.0",
     "react": "16.7.0",
-    "react-dom": "16.7.0"
+    "react-dom": "16.7.0",
```
Comment (nickcolley) [inline package.json]: I believe if the main library is now using React, we'll need to make it a 'dependency' not a 'devDependency'

**Comment by nickcolley** [comment]
Another question I have, will this solve issues for our vue friends?

**Comment by dakebl** [comment]
Hey @nickcolley , 

I'm sorry to hear that you've been feeling burnt out. Turning off the notifications is sometimes the self care we need. I hope you're feeling better now  🙌 

---
#### Framework Dependencies
In regards to the react dependency; I agree that it is heavy handed to bring in a library when we don't have to. I've instead created my own util for checking for react components.

That said, in order to get the `renderToString` function into the main logic, I had to make `vue` a dependency (as the function uses it as a peer dependency).

Generally speaking though, I think that it is ok for a testing lib to use framework utils/code to enable better testing DX, as long as the demand is there and it doesn't weigh down the end result. I imagine it will be rare for `jest-axe` to be packaged up in a build and will most often be a dev dependency anyway.

What do you think?

---

#### Updates to the PR:
- I've removed the react core dependency. I've created a util to check for react elements
-  I've added support for vue. I've added a util to check for vue elements and a lib for converting them to strings
- I've created a util for matching React Testing Library containers
- I've moved the dependencies around to match the main codes requirements.
- I've moved the tests into a `__tests__` directory and separated out the lib specific tests

---
#### Release Version
I believe this will be a feature release only.
I've made sure all of the existing tests pass, updated some to meet the new criteria, and added some new ones for Vue.

Let me know what you think ✌️

**Comment by rzane** [comment]
Casual onlooker here. I encountered issue #56, so I've been keeping an eye on this.

As a React user, I probably don't want to install Vue. As a Vue user, I probably don't want to install React. As a vanilla JS user, I probably don't want to install either.

I'm personally using `@testing-library/react`, so to get HTML, I can just do this:

```javascript
import { render } from "@testing-library/react";
await axe(render(<App />).container.innerHTML)
```

For people using `enzyme`, they can just do this:

```javascript
import { render } from "enzyme";
await axe(render(<App />).html());
```

For people using `@vue/test-utils`, they can just do this:

```javascript
import { mount } from "@vue/test-utils";
await axe(mount(App).html());
```

I appreciate the work you've put in @dakebl, but would it make sense if this library just accepted HTML (the common denominator)? Is it worth the additional complexity for this library to be able to accept React/Vue nodes?

**Comment by nickcolley** [comment]
I also really appreciate the time and care you've taken @dakebl to try and accommodate the feedback so far.

I am interested in the approach proposed by @rzane so we can avoid adding vue and react dependencies to this library entirely, based on your thinking so far do you think this alongside some good guidance would solve the issues?

**Comment by dakebl** [comment]
Hey @rzane,

Thanks for taking the time to take a look.
You make a very good point and I agree @nickcolley, documenting a better approach for each library/framework is much better than baking in the libraries/frameworks themselves.

The reason I started adding these was to allow me to use the maintained methods for library element detection, which I was needing in order to rectify the initial problem of duplicate IDs.
The issue stems from the code being attached to the DOM twice; once by a testing framework/library, and a second time by jest-axe. 

My solution was to detect if a library had been used and, instead of adding it a second time, render the document in `axeCore` directly. This, while solving the issue, creeped into me adding libs and complicating things unnecessarily. 

The issue as it stands is, how do we ensure that the DOM is clean before running the test.

I've experimented with a few things like cloning the `document` and attaching to the clone, passing a smaller scope to `axeCore.run`, etc, but I've yet to do it elegantly.
So far, the best working fix I've come up with is storing the `innerHTML` of the body, replacing it with the passed `html`, and reinstating the original `innerHTML` again after testing.

This downside of this approach was that it caused the `should not mutate the content of document.body permanently` test to begin failing, as the body was not the exact DOM element it generated and the test expected the returning body html to be a ref of the original.
I changed the `expect` statement from a `toBe` assertion to the less strict `toEqual` assertion, which then allows the tests to pass (as the body copy is equivalent, just not the same exact ref) but I'm not sure how I feel about this.

What do you think @nickcolley @rzane? I've pushed the code for your consideration.

**Comment by rzane** [comment]
All of the decisions you made above make sense to me (specifically, swapping the `innerHTML` and using `toEqual`). One other thing to consider is this though: if the element is already mounted in the document, there's no need to do the body-swapping. This scenario is what happens with enzyme.

I took a stab at handling that scenario here: https://github.com/rzane/jest-axe/blob/56-render-react-and-rtl-elements/index.js. I also moved some of this logic into a function called `mount`.

**Comment by dakebl** [comment]
@rzane Ha, you're right! That's a lot more elegant, thanks for creating the abstractions 🙌 
I've pulled in your changes and pushed them for @nickcolley to review.

**Comment by dakebl** [comment]
Added a `@testing-library/vue` test:

```js
import { render } from "@testing-library/react";
const { container } = render(<Image />);
const results = await axe(container);

expect(results).toHaveNoViolations()

cleanup()
```

File: README.md [hunk 3]
```diff
@@ -47,26 +47,70 @@ it('should demonstrate this matcher`s usage', async () => {
...
-  const html = ReactDOMServer.renderToString(
-    <img src='#' />
-  )
+  const results = await axe(<App/>)
+  expect(results).toHaveNoViolations()
+})
+```
 
-  const results = await axe(html)
+### With React Testing Library
```
Comment (nickcolley) [inline README.md]: Where possible I've tried to keep the test fixtures as close to the README snippets as possible.

Is it possible to keep them the same, the README examples seem simpler than what is being tested, are they correct?

**Comment by nickcolley** [comment]
I really like the new approach, very tidy code too.

If @rzane gives it the :+1: I think we can get this merged and released.

**Comment by dakebl** [comment]
That's great news @nickcolley, thanks! Looking forward to hearing @rzane's thoughts.
I've pushed an update to the README for you both to review that reflects my current thoughts of the recommended usage. Let me know what you both think 👍

File: index.js [hunk 4]
```diff
@@ -1,10 +1,40 @@
...
-
 const axeCore = require('axe-core')
 const merge = require('lodash.merge')
 const chalk = require('chalk')
 const { printReceived, matcherHint } = require('jest-matcher-utils')
 
+/**
+ * Converts a HTML string or HTML element to a mounted HTML element.
+ * @param {element | string} a HTML element or a HTML string
+ * @returns {[element, function]} a HTML element and a function to restore the document
```
Comment (rzane) [inline index.js]: `element` is not a valid JSDoc type. `Element` is a valid type, though.

File: README.md [hunk 5]
```diff
@@ -43,29 +43,101 @@ it('should demonstrate this matcher`s usage', async () => {
...
 
 ```javascript
+const React = require('react')
+const App = require('./app')
+
 const { axe, toHaveNoViolations } = require('jest-axe')
+expect.extend(toHaveNoViolations)
+
+it('should demonstrate this matcher`s usage with react', async () => {
+  const results = await axe(<App/>)
```
Comment (rzane) [inline README.md]: This example should probably needs to be updated.

File: README.md [hunk 6]
```diff
@@ -43,29 +43,101 @@ it('should demonstrate this matcher`s usage', async () => {
...
 
+const { mount } = require('enzyme')
+const { axe, toHaveNoViolations } = require('jest-axe')
 expect.extend(toHaveNoViolations)
 
+it('should demonstrate this matcher`s usage with enzyme', async () => {
+  const wrapper = mount(<App/>)
+  const results = await axe(container.getDOMNode())
+  
+  expect(results).toHaveNoViolations()
```
Comment (rzane) [inline README.md]: This example should say:

```javascript
  const wrapper = mount(<App/>, { attachTo: document.body })
  const results = await axe(container.getDOMNode())
  
  expect(results).toHaveNoViolations()
  wrapper.detach()
```

File: README.md [hunk 7]
```diff
@@ -43,29 +43,101 @@ it('should demonstrate this matcher`s usage', async () => {
...
+
+const { mount } = require('@vue/test-utils');
+const { axe, toHaveNoViolations } = require('jest-axe')
+expect.extend(toHaveNoViolations)
+
+it('should demonstrate this matcher`s usage with vue test utils', async () => {
+  const wrapper = mount(Image)
+  const results = await axe(wrapper.element)
 
   expect(results).toHaveNoViolations()
```
Comment (rzane) [inline README.md]:
```javascript
  const wrapper = mount(Image, { attachToDocument: true })
  const results = await axe(wrapper.element)
  expect(results).toHaveNoViolations()
  wrapper.destroy()
```

**Comment by citypaul** [comment]
Hey guys,

I'm on holiday at the moment so only just saw these updates. I'm unable to go through in any detail, but just wanted to say thank you to everyone involved for helping out here, and also I totally understand the burnout thing @nickcolley. I've got two young kids and find it hard to keep up with stuff at times. Mental health is essential. Just wanted to say thanks for this great library, and you should feel proud that people all over the world are using the work you have produced to make the web more accessible for more users.

Thanks everyone, but I'm off now for more holiday 🍻🏝️

File: README.md (hunk 7 above)
Comment (dakebl) [reply README.md]: We don't need to use the `attachToDocument` argument. This lib already takes care of this for us, and saves us having to clean up afterwards with `.destroy()`

File: README.md (hunk 6 above)
Comment (dakebl) [reply README.md]: We don't need to use the `attachT` argument. This lib already takes care of this for us, and saves us having to clean up afterwards with `.detach()`

File: README.md (hunk 5 above)
Comment (dakebl) [reply README.md]: What needs updating?

File: index.js (hunk 4 above)
Comment (dakebl) [reply index.js]: Ah, yes, sorry. I have reverted this.

File: __tests__/vuejs.test.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+const { axe, toHaveNoViolations } = require('../index')
+
+const Image = require('./Image.vue')
```
[the remark this answers is not in the document]
Comment (dakebl) [reply __tests__/vuejs.test.js]: Good call. Updated.

File: package.json [hunk 9]
```diff
@@ -30,14 +30,39 @@
     "lodash.merge": "^4.6.2"
   },
   "devDependencies": {
```
[the remark this answers is not in the document]
Comment (dakebl) [reply package.json]: Good call! Removed.

**Comment by rzane** [comment]
Hey @dakebl, it looks like node 6 is failing because `async/await` isn't supported in that version. Since Node 6 reached end-of-life in April, it's probably safe to remove that version from the `.travis.yml`.

**Comment by dakebl** [comment]
> Hey @dakebl, it looks like node 6 is failing because `async/await` isn't supported in that version. Since Node 6 reached end-of-life in April, it's probably safe to remove that version from the `.travis.yml`.

Fixed @rzane. What do you think about my comments on the README examples?

File: README.md (hunk 5 above)
Comment (rzane) [reply README.md]: The `axe` function does not accept a React component. It can only accept a string containing HTML or an HTML element. This example should probably use `react-dom`:

```javascript
import { render } from "react-dom";
render(<App />, document.body);
await axe(document.body);
```

File: README.md (hunk 6 above)
Comment (rzane) [reply README.md]: You're not wrong, but I think `jest-axe` should avoid modifying the underlying document whenever possible. I think it's probably better to let people's test frameworks handle that part. The fact that `jest-axe` will put HTML into the document should probably be a sort of "fallback plan".

Comment (dakebl) [reply README.md]: Adding the `attachTo: document.body` causing a large warning during to testing: 

```
Warning: render(): Rendering components directly into document.body is discouraged,
since its children are often manipulated by third-party scripts and browser extensions.
This may lead to subtle reconciliation issues. Try rendering into a container element
created for your app.
```

Should we omit the `attachTo`, given then error?
Or should we go with something like: 
```js
const testContainer = document.createElement('div')
document.body.appendChild(testContainer)

const element = React.createElement('img', { src: '#' })
const wrapper = mount(element, { attachTo: testContainer })
const results = await axe(wrapper.getDOMNode())
```

File: README.md (hunk 7 above)
Comment (dakebl) [reply README.md]: I've updated this one

File: README.md (hunk 5 above)
Comment (dakebl) [reply README.md]: I've updated this in the README

File: README.md (hunk 6 above)
Comment (rzane) [reply README.md]: Eek, yeah, that's annoying. It's *probably* fine to omit the `attachTo`, but I'm a little worried about the fact that when you replace the `innerHTML`, it won't restore the event handlers.

I guess since `enzyme` and `vue` aren't actually rendering into the document by default anyway, it's probably fine that jest-axe modifies the document.body in those cases.

Comment (dakebl) [reply README.md]: Yeah, I agree. It's not ideal but given the current way they are handling rendering, I think it's ok. I've updated the README accordingly.

File: README.md (hunk 3 above)
Comment (dakebl) [reply README.md]: I've updated all of the tests and README to match as closely as possible

**Comment by dakebl** [comment]
Are you happy to give this a quick approval @rzane?

**Comment by dakebl** [comment]
Are you happy to give this a quick approval @rzane? 🙏

**Comment by dakebl** [comment]
Are you happy to merge this now @nickcolley?

**Comment by nickcolley** [comment]
I've asked Kent to take a peak at this since he's opened a different PR, if he's busy and can't review this I'll get this merged.

**Comment by nickcolley** [comment]
OK thanks a bunch everyone for helping me with this, I'm going to try and clean this stuff up and get a release out.

**Comment by dakebl** [comment]
That's great @nickcolley! Thank you for merging it in 🙌 I'll keep an eye out for anything else I can help with in the future 👀
